chainstate/src/detail/chainstateref/mod.rs

```diff
@@ -987,6 +994,107 @@ impl<'a, S: BlockchainStorageRead, V: TransactionVerificationStrategy> Chainstat
         Ok(block_index)
     }
 
+    pub fn get_stake_pool_balances_at_heights(
+        &self,
+        pool_ids: &[PoolId],
+        min_height: BlockHeight,
+        max_height: BlockHeight,
+    ) -> Result<BTreeMap<BlockHeight, BTreeMap<PoolId, Amount>>, BlockError> {
+        let best_block_index =
+            self.get_best_block_index().map_err(BlockError::PropertyQueryError)?;
+        let best_block_height = best_block_index.block_height();
+
+        ensure!(
+            min_height <= max_height && max_height <= best_block_height,
+            BlockError::UnexpectedHeightRange(min_height, max_height)
+        );
+
+        // This will track ids of pools that have non-zero balance above the current height.
+        let mut pool_ids_that_had_balance = BTreeSet::new();
+        // If a pool has no balance at a certain height but it's known to have one
+        // at a bigger height, it means that it is created above that height,
+        // so there is no point in checking its balance below it.
+        // Such pool ids will be removed from the set; if the set becomes empty, we'll stop
+        // iterating, to avoid performing useless reorgs.
+        let mut pool_ids = pool_ids.iter().copied().collect::<BTreeSet<_>>();
+
+        let mut height_map = BTreeMap::new();
+
+        let max_height = if max_height == best_block_height {
+            let balances_at_tip = Self::collect_pool_balances(pool_ids.iter(), self)?;
+            if !balances_at_tip.is_empty() {
+                height_map.insert(best_block_height, balances_at_tip);
+            }
+
+            if max_height == min_height {
+                return Ok(height_map);
+            }
+
+            max_height.prev_height().expect("max_height can't be zero at this point")
+        } else {
+            max_height
+        };
+
+        let lowest_block_id = self
+            .get_existing_block_id_by_height(&min_height)
+            .map_err(BlockError::PropertyQueryError)?;
+
+        self.disconnect_tip_in_memory_until(
+            &lowest_block_id,
+            |disconnected_block_index, tx_verifier, _| -> Result<_, BlockError> {
+                let cur_height = disconnected_block_index
+                    .block_height()
+                    .prev_height()
+                    .expect("Genesis can't be disconnected");
+                assert!(cur_height >= min_height);
+
+                let pos_db = PoSAccountingDB::<_, TipStorageTag>::new(&self.db_tx);
+                let pos_delta =
+                    PoSAccountingDeltaRef::new(&pos_db, tx_verifier.accounting_delta_data());
```

Comment:
I think you don't need this at all. `TransactionVerifier` already implements `PoSAccountingView`

Reply:
> I think you don't need this at all. TransactionVerifier already implements PoSAccountingView

Indeed, thanks. I've reverted adding that `PoSAccountingDeltaRef` stuff.